graphblas/core/matrix.py

```diff
@@ -2805,6 +2805,119 @@ def power(self, n, op=semiring.plus_times):
+            v = values
+            clear_diag = accum is None and v._nvals != v._size
+        elif type(values) is Scalar:
+            is_scalar = True
+        else:
+            dtype = self.dtype if self.dtype._is_udt else None
+            try:
+                # Try to make it a Scalar
+                values = Scalar.from_value(values, dtype, is_cscalar=None, name="")
+                is_scalar = True
+            except (TypeError, ValueError):
+                try:
+                    # Else try to make it a numpy array
+                    values, dtype = values_to_numpy_buffer(values, dtype)
+                except Exception:
+                    self._expect_type(
+                        values,
+                        (Scalar, Vector, np.ndarray),
+                        within="setdiag",
+                        argname="values",
+                        extra_message="Literal scalars also accepted.",
+                    )
+                else:
+                    v = Vector.from_dense(values, dtype=dtype, **opts)
+
+        if is_scalar:
+            v = Vector.from_scalar(values, size, **opts)
+        elif v._size != size:
+            raise DimensionMismatch(
+                f"Dimensions not compatible for assigning length {v._size} Vector "
+                f"to {k}'th diagonal of Matrix with shape {self._nrows}x{self._ncols}."
+                f"The Vector should be size {size}."
+            )
+
+        if mask is not None:
+            mask = _check_mask(mask)
+            if mask.parent.ndim == 2:
+                if mask.parent.shape != self.shape:
+                    raise DimensionMismatch(
+                        "Matrix mask in setdiag is the wrong shape; "
+                        f"expected shape {self._nrows}x{self._ncols}, "
+                        f"got {mask.parent._nrows}x{mask.parent._ncols}"
+                    )
+                if mask.complement:
+                    mval = type(mask)(mask.parent.diag(k)).new(**opts)
+                    mask = mval.S
+                    M = mval.diag()
+                else:
+                    M = select.diag(mask.parent, k).new(**opts)
+            elif mask.parent._size != size:
+                raise DimensionMismatch(
+                    "Vector mask in setdiag is the wrong length; "
+                    f"expected size {size}, got size {mask.parent._size}."
+                )
+            else:
+                if mask.complement:
+                    mask = mask.new(**opts).S
+                M = mask.parent.diag()
+            if M.shape != self.shape:
+                M.resize(self._nrows, self._ncols)
+            mask = type(mask)(M)
+
+        if clear_diag:
+            self(mask=mask, **opts) << select.offdiag(self, k)
+
+        Diag = v.diag(k)
+        if Diag.shape != self.shape:
+            Diag.resize(self._nrows, self._ncols)
+        if mask is None:
+            mask = Diag.S
+        self(accum=accum, mask=mask, **opts) << Diag
```

Comment:
For posterity and those who browse this PR, these lines are the main recipe for setting the diagonal of a Matrix with a Vector. The rest of the function mostly deals with handing different input types and giving good error messages when necessary.